tencentcloud/data_source_tc_cbs_storages.go

```diff
@@ -168,50 +169,58 @@ func dataSourceTencentCloudCbsStoragesRead(d *schema.ResourceData, meta interfac
 	cbsService := CbsService{
 		client: meta.(*TencentCloudClient).apiV3Conn,
 	}
-	storages, err := cbsService.DescribeDisksByFilter(ctx, params)
-	if err != nil {
-		return err
-	}
 
-	storageList := make([]map[string]interface{}, 0, len(storages))
-	ids := make([]string, len(storages))
-	for _, storage := range storages {
-		mapping := map[string]interface{}{
-			"storage_id":        *storage.DiskId,
-			"storage_name":      *storage.DiskName,
-			"storage_usage":     *storage.DiskUsage,
-			"storage_type":      *storage.DiskType,
-			"availability_zone": *storage.Placement.Zone,
-			"project_id":        *storage.Placement.ProjectId,
-			"storage_size":      *storage.DiskSize,
-			"attached":          *storage.Attached,
-			"instance_id":       *storage.InstanceId,
-			"encrypt":           *storage.Encrypt,
-			"create_time":       *storage.CreateTime,
-			"status":            *storage.DiskState,
+	err := resource.Retry(readRetryTimeout, func() *resource.RetryError {
+		storages, e := cbsService.DescribeDisksByFilter(ctx, params)
+		if e != nil {
+			return retryError(e)
 		}
-		if storage.Tags != nil {
-			tags := make(map[string]interface{}, len(storage.Tags))
-			for _, t := range storage.Tags {
-				tags[*t.Key] = *t.Value
+		ids := make([]string, len(storages))
+		storageList := make([]map[string]interface{}, 0, len(storages))
+		for _, storage := range storages {
+			mapping := map[string]interface{}{
+				"storage_id":        *storage.DiskId,
+				"storage_name":      *storage.DiskName,
+				"storage_usage":     *storage.DiskUsage,
+				"storage_type":      *storage.DiskType,
+				"availability_zone": *storage.Placement.Zone,
+				"project_id":        *storage.Placement.ProjectId,
+				"storage_size":      *storage.DiskSize,
+				"attached":          *storage.Attached,
+				"instance_id":       *storage.InstanceId,
+				"encrypt":           *storage.Encrypt,
+				"create_time":       *storage.CreateTime,
+				"status":            *storage.DiskState,
 			}
-			mapping["tags"] = tags
+			if storage.Tags != nil {
+				tags := make(map[string]interface{}, len(storage.Tags))
+				for _, t := range storage.Tags {
+					tags[*t.Key] = *t.Value
+				}
+				mapping["tags"] = tags
+			}
+			storageList = append(storageList, mapping)
+			ids = append(ids, *storage.DiskId)
 		}
-		storageList = append(storageList, mapping)
-		ids = append(ids, *storage.DiskId)
-	}
 
-	d.SetId(dataResourceIdsHash(ids))
-	if err = d.Set("storage_list", storageList); err != nil {
-		log.Printf("[CRITAL]%s provider set storage list fail, reason:%s\n ", logId, err.Error())
-		return err
-	}
+		d.SetId(dataResourceIdsHash(ids))
```

Comment:
`d.SetId()` 是不是应该放到最后一步呢，因为下面 `d.Set("storage_list", storageList)` 可能会产生错误导致 list 设置不成功，而此时这个 data source create 应该视为失败，但是一旦调用了 `d.SetId()` terraform 就会视为创建成功。

https://www.terraform.io/docs/extend/writing-custom-providers.html
> it uses SetId, a built-in function, to set the ID of the resource to the address. The existence of a non-blank ID is what tells Terraform that a resource was created.

> Since the Create operation used SetId, Terraform believes the resource created successfully